sdk/identity/azure_identity/src/managed_identity_credential.rs

```diff
@@ -0,0 +1,498 @@
+    /// The [`TokenCredentialOptions`] to use for the credential.
+    pub credential_options: TokenCredentialOptions,
+
+    /// Specifies a user-assigned identity the credential should authenticate.
+    /// When `None`, the credential will authenticate a system-assigned identity, if any.
+    pub user_assigned_id: Option<UserAssignedId>,
+}
+
+impl ManagedIdentityCredential {
+    pub fn new(options: Option<ManagedIdentityCredentialOptions>) -> azure_core::Result<Arc<Self>> {
+        let options = options.unwrap_or_default();
+        let source = get_source();
+        let id = options
+            .user_assigned_id
+            .clone()
+            .map(|id| id.into())
+            .unwrap_or(ImdsId::SystemAssigned);
+
+        let credential: Arc<dyn TokenCredential> = match source {
+            ManagedIdentitySource::AppService => {
+                // App Service does accept resource IDs, however this crate's current implementation sends
+                // them in the wrong query parameter: https://github.com/Azure/azure-sdk-for-rust/issues/2407
+                if let ImdsId::MsiResId(_) = id {
+                    return Err(azure_core::Error::with_message(
+                        azure_core::error::ErrorKind::Credential,
+                        || {
+                            "User-assigned resource IDs aren't supported for App Service. Use a client or object ID instead.".to_string()
+                        },
+                    ));
+                }
+                AppServiceManagedIdentityCredential::new(id, options.credential_options)?
+            }
+            ManagedIdentitySource::Imds => {
+                VirtualMachineManagedIdentityCredential::new(id, options.credential_options)?
+            }
+            _ => {
+                return Err(azure_core::Error::with_message(
+                    azure_core::error::ErrorKind::Credential,
+                    || format!("{} managed identity isn't supported", source.as_str()),
+                ));
+            }
+        };
+
+        let mut message = format!(
+            "ManagedIdentityCredential will use {} managed identity",
+            source.as_str()
+        );
+        if let Some(user_assigned_id) = options.user_assigned_id {
+            message.push_str(&format!(" with {}", user_assigned_id));
+        }
+        trace!(message);
+
+        Ok(Arc::new(Self { credential }))
+    }
+}
+
+#[cfg_attr(target_arch = "wasm32", async_trait::async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait::async_trait)]
+impl TokenCredential for ManagedIdentityCredential {
+    async fn get_token(&self, scopes: &[&str]) -> azure_core::Result<AccessToken> {
+        self.credential.get_token(scopes).await
+    }
+}
+
+#[derive(Debug, Copy, Clone)]
+enum ManagedIdentitySource {
+    AzureArc,
+    AzureML,
+    AppService,
+    CloudShell,
+    Imds,
+    ServiceFabric,
+}
+
+impl ManagedIdentitySource {
+    pub fn as_str(&self) -> &str {
```

Comment:
```suggestion
    pub fn as_str(&self) -> &'static str {
```

May improve compiler optimizations.